From: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Date: Wed, 15 Sep 2021 15:59:44 +0200
Subject: timeout-for-ping

Origin: Ubuntu
Author: Brian Murray <brian.murray@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1705345, https://bugs.launchpad.net/bugs/1872159
Forwarded: no
Last-Update: 2020-10-21
---
 src/client/plymouth.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/client/plymouth.c b/src/client/plymouth.c
index be74ae7..11b8910 100644
--- a/src/client/plymouth.c
+++ b/src/client/plymouth.c
@@ -75,6 +75,13 @@ typedef struct
         char    *keys;
 } key_answer_state_t;
 
+ply_event_loop_timeout_handler_t timeout_handler(void *user_data)
+{
+        ply_event_loop_t *loop = user_data;
+        ply_trace ("ping: timed out waiting for plymouthd");
+        ply_event_loop_exit(loop, 1);
+}
+
 static char **
 split_string (const char *command,
               const char  delimiter)
@@ -1145,6 +1152,8 @@ main (int    argc,
                                              on_success,
                                              (ply_boot_client_response_handler_t)
                                              on_failure, &state);
+                ply_event_loop_watch_for_timeout (state.loop, 30.0,
+                                                  timeout_handler, state.loop);
         } else if (should_check_for_active_vt) {
                 ply_boot_client_ask_daemon_has_active_vt (state.client,
                                                           (ply_boot_client_response_handler_t)
